internal/oracle/identity.go

```diff
@@ -13,75 +13,163 @@ import (
+	var description string
+	if err := VerifyEntityIsSubset(entity, userEntity); err != nil {
+		description = err.Error()
+		approved = false
 		approve = 0
 	}
 
 	// Get block hash for tip - 4
 	blockHash, height, err := blockHandler.SigHash(ctx)
 	if err != nil {
-		return nil, 0, false, errors.Wrap(err, "get sig block hash")
+		return nil, 0, false, "", errors.Wrap(err, "get sig block hash")
 	}
 
-	sig, err := protocol.EntityXPubOracleSigHash(ctx, entity, xpub, &blockHash, approve)
+	sig, err := protocol.EntityXPubOracleSigHash(ctx, entity, xpub, blockHash, approve)
 	if err != nil {
-		return nil, 0, false, errors.Wrap(err, "generate signature")
+		return nil, 0, false, "", errors.Wrap(err, "generate signature")
 	}
 
-	return sig, height, true, nil
+	return sig, height, approved, description, nil
+}
+
+// CreateAdminCertificate creates an admin certificate for contract offers.
+// Returns:
+//   []byte - signature hash
+//   uint32 - block height of block hash included in signature hash
+//   bitcoin.Hash32 - block hash included in signature hash
+//   bool - true if approved
+func CreateAdminCertificate(ctx context.Context, dbConn *db.DB, isTest bool,
+	blockHandler *BlockHandler, xpubs bitcoin.ExtendedKeys, index uint32,
+	issuer actions.EntityField, entityContract bitcoin.RawAddress,
+	expiration uint64) ([]byte, uint32, bool, string, error) {
+
+	user, err := FetchUserByXPub(ctx, dbConn, xpubs)
+	if err != nil {
+		return nil, 0, false, "", errors.Wrap(err, "fetch user")
+	}
+
+	userEntity := &actions.EntityField{}
+	if err := proto.Unmarshal(user.Entity, userEntity); err != nil {
+		return nil, 0, false, "", errors.Wrap(err, "unmarshal user entity")
+	}
+
+	xpubData, err := FetchXPubByXPub(ctx, dbConn, xpubs)
+	if err != nil {
+		return nil, 0, false, "", errors.Wrap(err, "fetch xpub")
+	}
+
+	adminKey, err := xpubs.ChildKeys(index)
+	if err != nil {
+		return nil, 0, false, "", errors.Wrap(err, "generate address key")
+	}
+
+	adminAddress, err := adminKey.RawAddress(xpubData.RequiredSigners)
+	if err != nil {
+		return nil, 0, false, "", errors.Wrap(err, "generate address")
+	}
+
+	// Get block hash for tip - 4
+	blockHash, height, err := blockHandler.SigHash(ctx)
+	if err != nil {
+		return nil, 0, false, "", errors.Wrap(err, "get sig block hash")
+	}
+
+	approved := true
+	approve := uint8(1)
+	var description string
+	var entity interface{}
+	if entityContract.IsEmpty() {
+		entity = issuer
+
+		// Verify the entity matches that registered to the user.
+		if err := VerifyEntityIsSubset(&issuer, userEntity); err != nil {
+			description = err.Error()
+			approved = false
+			approve = 0
+		}
+	} else {
+		entity = entityContract
+
+		// Verify the contract belongs to the user.
+		cf, err := GetContractFormation(ctx, dbConn, entityContract, isTest)
+		if err != nil {
+			return nil, 0, false, "", errors.Wrap(err, "get contract formation")
+		}
+
+		if err := VerifyEntityIsSubset(cf.Issuer, userEntity); err != nil {
```

Comment:
These blocks can be improved.

The previous block, and the last half of this one are identical.